drivers/flash/spi_flash_sfdp.c

```diff
@@ -0,0 +1,558 @@
+			ptr = data->bfpt;
+			len = min(ARRAY_SIZE(data->bfpt), phdr[i].length);
+			break;
+		default:
+			continue;
+		}
+
+		spi_flash_read_sfdp(dev, get_phdr_addr(&phdr[i]),
+				    ptr, len * 4);
+	}
+
+	if (!data->bfp.density) {
+		return -ENODEV;
+	}
+
+	data->flash_size = data->bfp.density >> 3;
+	data->page_size = 1U << data->bfp.page_size_shift;
+
+	SYS_LOG_INF("SFDP v%d.%d %d KiB (%d page size)",
+		sfdp.major_ver, sfdp.minor_ver,
+		(data->flash_size + 1) >> 10,
+		data->page_size);
+
+	switch (data->bfp.addr_bytes) {
+	case 0: /* 3-bytes */
+	case 1: /* 3-bytes default (4-bytes supported) */
+		data->addr_len = 3;
+		break;
+	case 2: /* 4-bytes */
+		data->addr_len = 4;
+		break;
+	default:
+		return -ENODEV;
+	}
+
+	data->read_opcode = CMD_READ_MEMORY;
+	data->read_dummy = 0;
+
+	data->write_opcode = CMD_PROGRAM_PAGE;
+	data->write_dummy = 0;
+
+	return 0;
+}
+
+static int spi_flash_cmd(struct device *dev, u8_t cmd_id)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+	u8_t buf[1];
+	struct spi_buf cmd = {
+		.buf = buf,
+		.len = sizeof(buf),
+	};
+
+	buf[0] = cmd_id;
+
+	if (spi_transceive(&data->config, &cmd, 1, &cmd, 1) != 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int spi_flash_read(struct device *dev, off_t offset, void *ptr,
+			  size_t len)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+	u8_t hdr[5];
+	struct spi_buf txbufs[2] = {
+		{ .buf = hdr,  .len = 1 + data->addr_len },
+		{ .buf = NULL, .len = data->read_dummy + len },
+	};
+	struct spi_buf rxbufs[2] = {
+		{ .buf = NULL, .len = 1 + data->addr_len + data->read_dummy },
+		{ .buf = ptr,  .len = len },
+	};
+	int r, i;
+
+	if (offset < 0 || offset+len >= data->flash_size) {
+		return -ENODEV;
+	}
+
+	k_sem_take(&data->sem, K_FOREVER);
+	wait_for_flash_idle(dev);
+	i = 0;
+	hdr[i++] = data->read_opcode;
+	if (data->addr_len == 4) {
+		hdr[i++] = offset >> 24;
+	}
+	hdr[i++] = offset >> 16;
+	hdr[i++] = offset >> 8;
+	hdr[i++] = offset;
+	r = spi_transceive(&data->config, txbufs, ARRAY_SIZE(txbufs),
+			   rxbufs, ARRAY_SIZE(rxbufs));
+	k_sem_give(&data->sem);
+
+	return r;
+}
+
+static int spi_flash_write(struct device *dev, off_t offset,
+			      const void *ptr, size_t len)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+	u8_t hdr[5];
+	struct spi_buf txbufs[] = {
+		{ .buf = hdr, .len = 1 + data->addr_len },
+		{ .buf = NULL, .len = data->write_dummy },
+		{ .buf = (void *)ptr, .len = len },
+	};
+	int r, i;
+
+	SYS_LOG_DBG("write: @%x len %x, opcode %x (page size %d)",
+		offset, len,
+		data->write_opcode, data->page_size);
+
+	if (!data->write_opcode) {
+		return -ENOTSUP;
+	}
+
+	if (offset < 0 || offset+len >= data->flash_size
+	    || len > data->page_size) {
+		return -ENODEV;
+	}
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	wait_for_flash_idle(dev);
+
+	r = spi_flash_cmd(dev, CMD_WRITE_ENABLE);
+	if (r != 0) {
+		goto err;
+	}
+
+	i = 0;
+	hdr[i++] = data->write_opcode;
+	if (data->addr_len == 4) {
```

Comment:
add an empty line before (basicall: if assigned value is not used in the if condition: break a line).